lib/auth/grpcserver.go

```diff
@@ -840,6 +869,18 @@ func (g *GRPCServer) SetAccessRequestState(ctx context.Context, req *authpb.Requ
 }
 
 func (g *GRPCServer) SubmitAccessReview(ctx context.Context, review *types.AccessReviewSubmission) (*types.AccessRequestV3, error) {
+	// Prevent users from submitting access reviews with the "promoted" state.
+	// Promotion is only allowed by SubmitAccessReviewAllowPromotion API in Ent.
+	if review.Review.ProposedState.IsPromoted() {
+		return nil, trace.BadParameter("state promoted can be only set when promoting to access list")
+	}
+
+	return g.SubmitAccessReviewAllowPromotion(ctx, review)
+}
+
+// SubmitAccessReviewAllowPromotion is a method created only for internal use by Ent client, and it is not
+// exposed in our GRPC service.
+func (g *GRPCServer) SubmitAccessReviewAllowPromotion(ctx context.Context, review *types.AccessReviewSubmission) (*types.AccessRequestV3, error) {
```

Comment:
This seems strange to me.  I can't recall us ever needing to create fake methods on `GRPCServer` before.  Fells wrong.  Also, more importantly, this strikes me as a potential risk.  This method looks and feels like a method that is meant to be expose to the network (it's doing authz after all), but it actually circumvents a key access-control mechanism.  Even if it isn't actually exposed, this pattern strikes me as extremely prone to future human error.

Reply:
I do agree, but I don't see a good way of sharing code between `e` and `ServerWithRoles`. As a compromise, I extracted the validation logic to `func AuthorizeAccessReviewRequest`. I'm using it in both places. It looks cleaner and less hacky, but it requires the implementation of some part of it in two places.
I think it's an improvement, but I'm also open to better ideas.